### PR TITLE
Feature/code reviews

### PR DESCRIPTION
**Pull request by Rosscrowley**
I have added headers for the Code Review topic.

### CODE REVIEW

**Review by Paddyjh — CHANGES_REQUESTED** [review]
I think we should hold off on merging until we have content we are happy with for each of the topics.

**Comment by Isaac811** [comment]
Fair play. I agree


Best regards

Jiayu Liu


Patrick Haughey ***@***.***>于2024年3月20日 周三16:45写道：

**Review by Isaac811 — APPROVED** [review]

File: Handbook/QA Handbook.md [hunk 1]
```diff
@@ -1,9 +1,82 @@
-# Quality Assurance Handbook
+# **Quality Assurance Handbook**
```
Comment (Paddyjh) [inline Handbook/QA Handbook.md]: I think we should remove the bold syntax; since headers are already bold, the "**" isn't needed. Removing this should make it easier when merging into main

File: Handbook/QA Handbook.md [hunk 2]
```diff
@@ -1,9 +1,82 @@
...
 
 ## Code Reviews
 
+### What Are Code Reviews?
+ 
+ - A code review is the task of someone who is not the author reviewing a piece of code before it is merged to the main codebase 
+ - A code review is an essential practice to ensure that the code is high quality, aligns with predetermined standards, contains no mistakes
+ - Many companies practice different processes of how to carry out code reviews. Different processes depend on a number of factors such as, the level of experience of developers, specific project or regulatory requirements,and the complexity and lifecycle stage of the project, etc. Below are some different approaches:
+    - Checklist-Based Reviews - where reviewers follow a predetermined list of checks to follow while reviewing code.
+    - Pair Programming - where a developer codes and gets real-time feedback (code review) on their code.
```
Comment (Paddyjh) [inline Handbook/QA Handbook.md]: I think we could clarify that this involves two coders programming together, which is where the real-time feedback comes from. Although maybe this is implied enough by the title?

File: Handbook/QA Handbook.md [hunk 3]
```diff
@@ -1,9 +1,82 @@
...
+- For Reviewers
+    - The reviewer should be experienced enough, so that they know what they are looking for such as design, logic, and bugs
+    - Adhere to a predefined checklist that define the goals and standards of the review process
+    - Code reviews should not be rushed to avoid code smells entering the main codebase. Limiting code review sessions or setting a maximum line of code (LOC) per hour can ensure that the reviewer does not lose focus
+    - Feedback should be constructive. Providing constructive feedback helps to enhance understanding and learning rather then incite competition or conflict.
+- For Reviewees
+    - Reviewee must be receptive to feedback. They must have a willingness to learn and understand criticism is not personal.
+    - Code must be reviewed by the reviewee before requesting a code review, to save the reviewers from having to respond to miscellaneous errors such as spelling mistakes.
+    - Seek and give clarity when needed. Be prepared to offer clarity on code you have produced and seek clarity when feedback is unclear.
+    - Small, incremental code changes should be the aim as code with many lines is more straneous to review.
```
Comment (Paddyjh) [inline Handbook/QA Handbook.md]: Spelling mistake: "straneous" should be "strenuous"

File: Handbook/QA Handbook.md [hunk 4]
```diff
@@ -1,9 +1,82 @@
...
+
+![Googles Code Review Flow](../Image_Folder/GoogleCRFlow.png)
+
+### Best Practices for Conducting Effective Code Reviews
+
+- Although there are many different approaches to conducting a code review, there are some key practices that should be considered for the code review to be effective. 
+- As well as key practices, included in this section are some practices that we feel are beneficial to fostering a positive and effective feedback 
+
+- For Reviewers
+    - The reviewer should be experienced enough, so that they know what they are looking for such as design, logic, and bugs
```
Comment (Paddyjh) [inline Handbook/QA Handbook.md]: To make this sentence flow better, we could rephrase it to something like "... what they are looking for, such as issues relating to design, logic and bugs".

File: Handbook/QA Handbook.md [hunk 5]
```diff
@@ -1,9 +1,82 @@
...
+
+    ![Worthwhileness vs Conflict Potential Diagram](../Image_Folder/WorthVSConflict.png)
+
+- There are approaches both reviewers and reviewees can take to avoid conflict and encourage collaboration
+    - Reviewers
+        - Use collaborative language and avoid using the word "You". For example, a better way of saying "Your code is inefficient" would be "How could we optimize this for better performance".
+        - Avoid leaving many comments where possible as this can demotivate the reviewer, instead prioritise essential comments, which brings the code up to a level of acceptance.
+        - As well as leaving constructive feedback, always attempt to leave a positive comment on something you felt was done well. This can boost the confidence of the reviewee.
+    - Reviewee
+        - Seems simple but ensure you reply to comments left by a reviewer as well as thanking them for taking the time to review your code.
```
Comment (Paddyjh) [inline Handbook/QA Handbook.md]: Maybe changing the start of this to "It seems simple ..." could make this flow better. Alternatively we could just Start the sentence at "Ensure you reply ..."

**Review by Paddyjh — CHANGES_REQUESTED** [review]
Left a few suggestions and corrections. Any questions just lemme know !

File: Handbook/QA Handbook.md (hunk 1 above)
Comment (Rosscrowley) [reply Handbook/QA Handbook.md]: I agree @Paddyjh,  will make that change, thanks.

File: Handbook/QA Handbook.md (hunk 2 above)
Comment (Rosscrowley) [reply Handbook/QA Handbook.md]: I agree, I will alter the sentence slightly to make it clearer. Thanks.

File: Handbook/QA Handbook.md (hunk 4 above)
Comment (Rosscrowley) [reply Handbook/QA Handbook.md]: Agreed, will reword the sentence.

File: Handbook/QA Handbook.md [hunk 6]
```diff
@@ -6,4 +6,77 @@
...
+- Although there are many different approaches to conducting a code review, there are some key practices that should be considered for the code review to be effective. 
+- As well as key practices, included in this section are some practices that we feel are beneficial to fostering a positive and effective feedback 
+
+- For Reviewers
+    - The reviewer should possess enough experience to effectively identify key aspects such as design, logic, and bugs.
+    - Adhere to a predefined checklist that define the goals and standards of the review process
+    - Code reviews should not be rushed to avoid code smells entering the main codebase. Limiting code review sessions or setting a maximum line of code (LOC) per hour can ensure that the reviewer does not lose focus
+    - Feedback should be constructive. Providing constructive feedback helps to enhance understanding and learning rather then incite competition or conflict.
+- For Reviewees
+    - Reviewee must be receptive to feedback. They must have a willingness to learn and understand criticism is not personal.
```
Comment (Isaac811) [inline Handbook/QA Handbook.md:31]: Maybe we could change this to “code must be reviewed and tested by the reviewee” to ensure code is also fully tested before a code review.

File: Handbook/QA Handbook.md [hunk 7]
```diff
@@ -6,4 +6,77 @@
...
 
+### What Are Code Reviews?
+ 
+ - A code review is the task of someone who is not the author reviewing a piece of code before it is merged to the main codebase 
+ - A code review is an essential practice to ensure that the code is high quality, aligns with predetermined standards, contains no mistakes
+ - Many companies practice different processes of how to carry out code reviews. Different processes depend on a number of factors such as, the level of experience of developers, specific project or regulatory requirements,and the complexity and lifecycle stage of the project, etc. Below are some different approaches:
+    - Checklist-Based Reviews - where reviewers follow a predetermined list of checks to follow while reviewing code.
+    - Pair Programming - involves two developers, where a developer codes and gets real-time feedback off another developer (code review) on their code.
+    - Pre-Commit Reviews - This is where code is reviewed before it is merged into the main branch.
+- Google mandates code reviews and and follows the Pre-commit review. Below is a diagram of their Code Review Flow.
```
Comment (Isaac811) [inline Handbook/QA Handbook.md]: Remove the double “and” in this sentence.

**Review by Isaac811 — CHANGES_REQUESTED** [review]
Let me know if you have a questoins , left 2 comments there. Thanks!

File: Handbook/QA Handbook.md (hunk 6 above)
Comment (Rosscrowley) [reply Handbook/QA Handbook.md:31]: I will add that in. thanks Isaac!

File: Handbook/QA Handbook.md (hunk 7 above)
Comment (Rosscrowley) [reply Handbook/QA Handbook.md]: Thanks for spotting this Isaac!

File: Handbook/QA Handbook.md (hunk 6 above)
Comment (Rosscrowley) [reply Handbook/QA Handbook.md:31]: updated in latest commit

File: Handbook/QA Handbook.md (hunk 7 above)
Comment (Rosscrowley) [reply Handbook/QA Handbook.md]: updated in latest commit

File: Handbook/QA Handbook.md (hunk 6 above)
Comment (Isaac811) [reply Handbook/QA Handbook.md:31]: The Reviewers might spell wrong?

Comment (Rosscrowley) [reply Handbook/QA Handbook.md:31]: Apologies, I see my mistake, I will correct now! Thanks.

Comment (Rosscrowley) [reply Handbook/QA Handbook.md:31]: Corrected spelling mistake in latest commit!

**Review by Isaac811 — APPROVED** [review]

**Review by Paddyjh — CHANGES_REQUESTED** [review]
Everything looks good but i think we can introduce a standard height limit of 500px for diagrams to better organize the overall structure and readability of the document. 

This can be done by using a html image tag and specifying the height rather than the way we are currently doing it. I have attached an example below :

img src="name_of_image.png" height="500"

What do yous think ?

**Comment by Rosscrowley** [comment]
> Everything looks good but i think we can introduce a standard height limit of 500px for diagrams to better organize the overall structure and readability of the document.
> 
> This can be done by using a html image tag and specifying the height rather than the way we are currently doing it. I have attached an example below :
> 
> img src="name_of_image.png" height="500"
> 
> What do yous think ?

@Paddyjh I agree, I will reformat the images. Thanks!

File: Handbook/QA Handbook.md [hunk 8]
```diff
@@ -6,4 +6,77 @@
 
 ## Code Reviews
 
+### What Are Code Reviews?
+ 
+ - A code review is the task of someone who is not the author reviewing a piece of code before it is merged to the main codebase
```
Comment (Paddyjh) [inline Handbook/QA Handbook.md]: I think we could rephrase this a bit more clearly

File: Handbook/QA Handbook.md [hunk 9]
```diff
@@ -6,4 +6,77 @@
 
 ## Code Reviews
 
+### What Are Code Reviews?
+ 
+ - A code review is the task of someone who is not the author reviewing a piece of code before it is merged to the main codebase 
+ - A code review is an essential practice to ensure that the code is high quality, aligns with predetermined standards, contains no mistakes
```
Comment (Paddyjh) [inline Handbook/QA Handbook.md]: small thing, but maybe we could change this to " and contains no mistakes."

File: Handbook/QA Handbook.md [hunk 10]
```diff
@@ -6,4 +6,77 @@
...
+    - Checklist-Based Reviews - where reviewers follow a predetermined list of checks to follow while reviewing code.
+    - Pair Programming - involves two developers, where a developer codes and gets real-time feedback off another developer (code review) on their code.
+    - Pre-Commit Reviews - This is where code is reviewed before it is merged into the main branch.
+- Google mandates code reviews and follows the Pre-commit review. Below is a diagram of their Code Review Flow.
+
+<img src="../Image_Folder/GoogleCRFlow.png" height="500" width="500"> 
+
+### Best Practices for Conducting Effective Code Reviews
+
+- Although there are many different approaches to conducting a code review, there are some key practices that should be considered for the code review to be effective.
```
Comment (Paddyjh) [inline Handbook/QA Handbook.md]: Could merge these two sentences into one ?

**Review by Paddyjh — CHANGES_REQUESTED** [review]
Just some small changes before we can merge, thanks

File: Handbook/QA Handbook.md (hunk 8 above)
Comment (Rosscrowley) [reply Handbook/QA Handbook.md]: Sounds good, how about, "A code review is the process of getting feedback and approval from other developers on submitted code before it is merged into the main codebase."

Comment (Paddyjh) [reply Handbook/QA Handbook.md]: That's perfect, thanks!

**Review by Paddyjh — APPROVED** [review]
Looks great, thanks for all the work on this guys!
